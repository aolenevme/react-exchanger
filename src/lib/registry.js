import store from "../store.js";

function updateIn(currentStore, payload) {
    const [path, value] = payload;

    path.reduce((acc, key, index) => {
        if (index === path.length - 1) {
            acc[key] = value;
        } else {
            return acc[key];
        }
    }, currentStore);
}

const storeEventsRegistry = {};

function dispatch(id, payload) {
    if (!id) {
        console.error("Enable to dispatch a store event without its id");

        return;
    }

    if (!storeEventsRegistry[id]) {
        console.error(
            `There isn\`t a registered handler for this store event id: ${id}`
        );

        return;
    }

    updateIn(store, storeEventsRegistry[id](store, payload));
}

function regEventStore(id, storeEvent) {
    if (!id) {
        console.error("Unable to register the event without its id");
    }

    if (!storeEvent) {
        console.error("Unable to register the store event without its handler");
    }

    if (storeEventsRegistry[id]) {
        console.error(
            `Unable to register the store event because it is already 
            registered: ${storeEventsRegistry[id]}`
        );
    }

    storeEventsRegistry[id] = storeEvent;
}

export {dispatch, regEventStore};
