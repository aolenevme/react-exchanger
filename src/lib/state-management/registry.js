import store from "../../store/store.js";

const storeEventsRegistry = {};

function dispatch(id, payload) {
    if (!id) {
        console.error("Enable to dispatch a store event without its id");

        return;
    }

    if (!storeEventsRegistry[id]) {
        console.error(
            `There isn\`t a registered handler for this store event id: ${id}`
        );

        return;
    }

    updateIn(store, storeEventsRegistry[id](store, payload));
}

function updateIn(currentStore, payload) {
    const [path, value] = payload;

    path.reduce((acc, key, index) => {
        if (index === path.length - 1) {
            acc[key] = value;
        } else {
            return acc[key];
        }
    }, currentStore);
}

function regEventStore(id, storeEvent) {
    if (!id) {
        console.error("Unable to register the event without its id");
    }

    if (!storeEvent) {
        console.error("Unable to register the store event without its handler");
    }

    if (storeEventsRegistry[id]) {
        console.error(
            `Unable to register the store event because it is already 
            registered: ${storeEventsRegistry[id]}`
        );
    }

    storeEventsRegistry[id] = storeEvent;
}

export {dispatch, regEventStore};
