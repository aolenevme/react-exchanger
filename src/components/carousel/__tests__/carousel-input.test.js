import React from "react";
import renderer from "react-test-renderer";

import * as registry from "../../../lib/state-management/registry.js";
import MUTATE_STORE from "../../../events/mutate-store.js";
import CarouselInput from "../carousel-input.js";
import checkStoreMutations from "../../../lib/tests/check-store-mutations.js";

describe("<CarouselInput />", () => {
    describe("rendering", () => {
        it("is rendered correctly", () => {
            expect(renderer
                .create(<CarouselInput />).toJSON())
                .toMatchSnapshot();

            expect(renderer
                .create(<CarouselInput isDisabled inputSign="£" value={12} />).toJSON())
                .toMatchSnapshot();

            expect(renderer
                .create(<CarouselInput isDisabled inputSign="£" value="0.00" />).toJSON())
                .toMatchSnapshot();
        });
    });

    describe("events", () => {
        beforeEach(() => {
            registry.dispatch = jest.fn();
        });

        afterEach(() => {
            registry.dispatch.mockClear();
        });

        const bigTestBalance = 500;

        it("updates store with a new digit value", () => {
            const testInputEvent = {target: {value: 123}};

            // eslint-disable-next-line no-magic-numbers,max-len
            renderer.create(<CarouselInput balance={bigTestBalance}/>).toJSON().children[2].props.onInput(testInputEvent);

            checkStoreMutations(0, MUTATE_STORE, [["exchangeAmount"], testInputEvent.target.value]);
        });

        it("updates the html input element with a previous input value and resets store in an emergency case", () => {
            const prevInputValue = 123;
            const testInputEvent = {target: {value: ""}};

            renderer.create(<CarouselInput
                balance={bigTestBalance}
                value={prevInputValue} />)
                // eslint-disable-next-line no-magic-numbers
                .toJSON().children[2].props.onInput(testInputEvent);

            checkStoreMutations(0, MUTATE_STORE, [["exchangeAmount"], ""]);
            expect(testInputEvent.target.value).toEqual(prevInputValue);
        });

        it("doesn`t allow to use floats with more than 2 digits after the dot", () => {
            const prevInputValue = 123.12;
            const testInputEvent = {target: {value: 123.123}};

            // eslint-disable-next-line no-magic-numbers,max-len
            renderer.create(<CarouselInput balance={bigTestBalance} value={prevInputValue} />).toJSON().children[2].props.onInput(testInputEvent);

            checkStoreMutations(0, MUTATE_STORE, [["exchangeAmount"], prevInputValue]);
        });

        it("doesn`t allow add more than there is on the balance", () => {
            const lowTestBalance = 123.13;
            const prevInputValue = 123.12;
            const testInputEvent = {target: {value: 123.14}};

            // eslint-disable-next-line no-magic-numbers,max-len
            renderer.create(<CarouselInput balance={lowTestBalance} value={prevInputValue} />).toJSON().children[2].props.onInput(testInputEvent);

            checkStoreMutations(0, MUTATE_STORE, [["exchangeAmount"], prevInputValue]);
        });
    });
});
