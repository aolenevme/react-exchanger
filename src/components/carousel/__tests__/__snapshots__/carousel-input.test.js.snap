// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarouselInput /> rendering is rendered correctly 1`] = `
.c0 {
  font-size: 3rem;
  line-height: 3rem;
  color: #f3f6ff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  padding: 0;
  min-width: 3rem;
  height: 3rem;
  border: 0 solid transparent;
  border-right: 1px solid transparent;
  background-color: transparent;
  outline: none;
  cursor: pointer;
}

.c0:focus-within {
  -webkit-animation: caret-pulse 1.5s cubic-bezier(.215,.61,.355,1) forwards infinite;
  animation: caret-pulse 1.5s cubic-bezier(.215,.61,.355,1) forwards infinite;
}

.c1 {
  margin: 0;
  padding: 0;
  width: 0;
  height: 0;
  border: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c1::-webkit-outer-spin-button,
.c1::-webkit-inner-spin-button {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c1:invalid {
  box-shadow: none;
}

<label
  className="main-text__MainText-c4b6rt-0 c0"
  tabIndex="-1"
>
  
  0
  <input
    className="c1"
    disabled={false}
    onChange={[Function]}
    onInput={[Function]}
    tabIndex="-1"
    type="number"
    value={0}
  />
</label>
`;

exports[`<CarouselInput /> rendering is rendered correctly 2`] = `
.c0 {
  font-size: 3rem;
  line-height: 3rem;
  color: #f3f6ff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  padding: 0;
  min-width: 3rem;
  height: 3rem;
  border: 0 solid transparent;
  border-right: 1px solid transparent;
  background-color: transparent;
  outline: none;
  cursor: initial;
}

.c0:focus-within {
  -webkit-animation: caret-pulse 1.5s cubic-bezier(.215,.61,.355,1) forwards infinite;
  animation: caret-pulse 1.5s cubic-bezier(.215,.61,.355,1) forwards infinite;
}

.c1 {
  margin: 0;
  padding: 0;
  width: 0;
  height: 0;
  border: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c1::-webkit-outer-spin-button,
.c1::-webkit-inner-spin-button {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c1:invalid {
  box-shadow: none;
}

<label
  className="main-text__MainText-c4b6rt-0 c0"
  tabIndex="-1"
>
  
  12
  <input
    className="c1"
    disabled={true}
    onChange={[Function]}
    onInput={[Function]}
    tabIndex="-1"
    type="number"
    value={12}
  />
</label>
`;

exports[`<CarouselInput /> rendering is rendered correctly 3`] = `
.c0 {
  font-size: 3rem;
  line-height: 3rem;
  color: #f3f6ff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  padding: 0;
  min-width: 3rem;
  height: 3rem;
  border: 0 solid transparent;
  border-right: 1px solid transparent;
  background-color: transparent;
  outline: none;
  cursor: initial;
}

.c0:focus-within {
  -webkit-animation: caret-pulse 1.5s cubic-bezier(.215,.61,.355,1) forwards infinite;
  animation: caret-pulse 1.5s cubic-bezier(.215,.61,.355,1) forwards infinite;
}

.c1 {
  margin: 0;
  padding: 0;
  width: 0;
  height: 0;
  border: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c1::-webkit-outer-spin-button,
.c1::-webkit-inner-spin-button {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c1:invalid {
  box-shadow: none;
}

<label
  className="main-text__MainText-c4b6rt-0 c0"
  tabIndex="-1"
>
  
  0
  <input
    className="c1"
    disabled={true}
    onChange={[Function]}
    onInput={[Function]}
    tabIndex="-1"
    type="number"
    value={0}
  />
</label>
`;
